x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/TreeNode.java

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.categorization;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.apache.lucene.util.Accountable;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.search.aggregations.AggregationExecutionException;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.stream.Collectors;
+
+import static org.apache.lucene.util.RamUsageEstimator.NUM_BYTES_OBJECT_REF;
+import static org.apache.lucene.util.RamUsageEstimator.sizeOfCollection;
+import static org.apache.lucene.util.RamUsageEstimator.sizeOfMap;
+import static org.elasticsearch.xpack.ml.aggs.categorization.CategorizationTokenTree.WILD_CARD;
+
+/**
+ * Tree node classes for the categorization token tree.
+ *
+ * Two major node types exist:
+ *  - Inner: which are nodes that have children token nodes
+ *  - Leaf: Which collection multiple {@link LogGroup} based on similarity restrictions
+ */
+abstract class TreeNode implements Accountable {
+
+    private static final Logger LOGGER = LogManager.getLogger(TreeNode.class);
+
+    private long count;
+
+    TreeNode(long count) {
+        this.count = count;
+    }
+
+    abstract void mergeWith(TreeNode otherNode);
+
+    abstract boolean isLeaf();
+
+    final void incCount(long count) {
+        this.count += count;
+    }
+
+    final long getCount() {
+        return count;
+    }
+
+    // TODO add option for calculating the cost of adding the new group
+    abstract LogGroup addLog(BytesRef[] logTokens, long docCount, TreeNodeFactory treeNodeFactory);
```

Comment:
I wonder if folks are going to be more interested in what they can keep for a certain amount of memory rather than what they can keep for some depth.....

Reply:
> I wonder if folks are going to be more interested in what they can keep for a certain amount of memory rather than what they can keep for some depth.....

An option that limits categories based on memory would be an interesting one. But, this PR is pretty beefy already :D.